docs/src/auth.md

```diff
@@ -571,18 +571,41 @@ test('admin and user', async ({ adminPage, userPage }) => {
 Reusing authenticated state covers [cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies) and [local storage](https://developer.mozilla.org/en-US/docs/Web/API/Storage) based authentication. Rarely, [session storage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage) is used for storing information associated with the signed-in state. Session storage is specific to a particular domain and is not persisted across page loads. Playwright does not provide API to persist session storage, but the following snippet can be used to save/load session storage.
 
 ```js
-// Get session storage and store as env variable
-const sessionStorage = await page.evaluate(() => JSON.stringify(sessionStorage));
-fs.writeFileSync('playwright/.auth/session.json', JSON.stringify(sessionStorage), 'utf-8');
+// For reading/writing session storage
+import fs from 'fs';
 
-// Set session storage in a new context
-const sessionStorage = JSON.parse(fs.readFileSync('playwright/.auth/session.json', 'utf-8'));
-await context.addInitScript(storage => {
-  if (window.location.hostname === 'example.com') {
-    for (const [key, value] of Object.entries(storage))
-      window.sessionStorage.setItem(key, value);
+/**
+ * Get session storage and store to a JSON file as env variable.
+ * This function should be invoked in the auth setup file.
+ * If you don't have multiple roles, you can set the file path fixed in the function.
+ */
+export async function writeSessionStorageToJSON(page, filePath) {
+  const sessionStorage = await page.evaluate(() => JSON.stringify(window.sessionStorage));
+  // we should use `writeFile` here instead of `writeFileSync`,
+  // especially if the file doesn't exist.
+  fs.writeFile(filePath, sessionStorage, { encoding: 'utf-8' }, err => {
```

Comment:
This is a bug, you are exiting writeSessionStorageToJSON before you wrote the file, you should use fs.promises.writeFile instead.

Reply:
hmmm alright, good catch!